File: Go/Lesson_067_reverse_vowels_of_a_string.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 067 -- Reverse Vowels of a String
// Category   : Two Pointers
// Difficulty : Easy
// Study Plan : Day 34
// =============================================================
//
// QUESTION:
//   Reverse only the vowels of a string. Vowels: a,e,i,o,u (and uppercase).
//
//   Example: hello -> holle
// =============================================================

package main
import "fmt"
func reverseVowels(s string) string {
    v := "aeiouAEIOU"; a := []byte(s); l,r := 0,len(a)-1
    isV := func(c byte) bool { for i:=0;i<len(v);i++ { if v[i]==c { return true } }; return false }
    for l<r {
        for l<r && !isV(a[l]) { l++ }
        for l<r && !isV(a[r]) { r-- }
        a[l],a[r] = a[r],a[l]; l++; r--
    }
    return string(a)
}
func main(){ fmt.Println(reverseVowels("hello"), reverseVowels("leetcode")) }

File: CPP/Lesson_067_reverse_vowels_of_a_string.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 067 -- Reverse Vowels of a String
// Category   : Two Pointers
// Difficulty : Easy
// Study Plan : Day 34
// =============================================================
//
// QUESTION:
//   Reverse only the vowels of a string. Vowels: a,e,i,o,u (and uppercase).
//
//   Example: hello -> holle
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
class Solution { public:
    string reverseVowels(string s){
        string v="aeiouAEIOU"; int l=0,r=s.size()-1;
        while (l<r){ while (l<r && v.find(s[l])==string::npos) l++; while (l<r && v.find(s[r])==string::npos) r--;
            swap(s[l],s[r]); l++; r--; }
        return s;
    }
};
int main(){ cout<<Solution().reverseVowels("hello")<<" "<<Solution().reverseVowels("leetcode")<<endl; }
